lms/djangoapps/course_blocks/transformers/library_content.py

```diff
@@ -98,13 +102,14 @@ def transform_block_filters(self, usage_info, block_structure):
                 # Save back any changes
                 if any(block_keys[changed] for changed in ('invalid', 'overlimit', 'added')):
                     state_dict['selected'] = list(selected)
-                    StudentModule.objects.update_or_create(
-                        student=usage_info.user,
-                        course_id=usage_info.course_key,
-                        module_state_key=block_key,
-                        defaults={
-                            'state': json.dumps(state_dict),
-                        },
+                    if not isinstance(usage_info.user, AnonymousUser):
+                        StudentModule.objects.update_or_create(
```

Comment:
This thing with this approach is that it is going to cause `make_selection` to return a different set of XBlocks on every view. To keep it stable may make more sense to add a mode `first` which returns the first `max_count` items from `library_children`.

And we can move the check to the first line: `if not isinstance(usage_info.user, AnonymousUser) and any(block_keys[changed] for changed in ('invalid', 'overlimit', 'added')):`

Reply:
Sorry I don't think I quite understood. We don't want it to return a different set of Xblocks on every view? I was under the impression that since we select `Randomized Content Block` while creating the course, the expected behaviour is that it should return random block(s) from the library. Is that not the case?

Reply:
My understanding is that the first time a learner accesses the XBlock a random set of children is calculated and stored in `state[selected]`. And they continue to see those same children in subsequence views (unless the children or the max_count values change where the list of children is updated).

Having said so I just realize that we do not have a `public_view` for the library xblock (the learner will be seeing the "Please enroll to view" message) and since we do not need to support it either for now lets just move the check to the outer if and let `make_selection` return a random set every time.

Nit: the convention is to use `[is_authenticated](https://docs.djangoproject.com/en/2.1/ref/contrib/auth/#django.contrib.auth.models.User.is_authenticated)` instead of the type of user objects.

Reply:
Thanks for the clarification. I've moved the check to the outer if.